axon-server-connector/src/main/java/org/axonframework/axonserver/connector/event/axon/EventBuffer.java

```diff
@@ -147,36 +147,51 @@ public void blacklist(TrackedEventMessage<?> trackedEventMessage) {
 
     @Override
     public boolean hasNextAvailable(int timeout, TimeUnit timeUnit) {
+        checkExceptionState();
         long deadline = System.currentTimeMillis() + timeUnit.toMillis(timeout);
         try {
             do {
                 long waitTime = deadline - System.currentTimeMillis();
                 waitForData(waitTime);
-            } while (peekEvent == null && System.currentTimeMillis() < deadline && !eventStream.hasNext());
+            } while (peekEvent == null && System.currentTimeMillis() < deadline);
 
-            return peekEvent != null || eventStream.hasNext();
+            return peekEvent != null;
         } catch (InterruptedException e) {
-            logger.warn("Consumer thread was interrupted. Returning thread to event processor.", e);
+            logger.warn("Event consumer thread was interrupted. Returning thread to event processor.", e);
             Thread.currentThread().interrupt();
             return false;
         }
     }
 
-    private void waitForData(long timeout) throws InterruptedException {
-        // a quick check before acquiring the lock
-        if (delegate.peek() != null) {
+    private void checkExceptionState() {
+        if (delegate.isClosed()) {
+            throw new AxonServerException(ErrorCode.OTHER.errorCode(),
+                                          "The Event Stream has been closed, so no further events can be retrieved");
+        }
+    }
+
+    private void waitForData(long waitTime) throws InterruptedException {
+        // Quick check before acquiring the lock.
+        if (peek().isPresent()) {
             return;
         }
-        if (timeout > 0) {
-            lock.lock();
-            try {
-                // check again for concurrency reasons
-                if (delegate.peek() == null) {
-                    dataAvailable.await(Math.min(DEFAULT_POLLING_TIME_MILLIS, timeout), TimeUnit.MILLISECONDS);
-                }
-            } finally {
-                lock.unlock();
+        // No use spending lock/await work for zero wait time.
+        if (waitTime <= 0) {
```

Comment:
can we merge this `if` with the previous one?
`if (peek().isPresent() || waitTime <= 0) { return;}`

Reply:
Sure, why not.